### PR TITLE
Projective transform for equivariant imaging

### PR DESCRIPTION
**Pull request by Andrewwango**
This PR introduces the projective transformation, or homography, an image transformation available for self-supervised learning using equivariant imaging. The transformation is implemented at `deepinv.transform.Homography`. 

I've also implemented a few subgroups of transformations, including `deepinv.transform.projective.Affine`, `deepinv.transform.projective.Similarity` and `deepinv.transform.projective.Euclidean` transformations, which may also be useful in some use-cases.

This was introduced in our recent paper [Perspective-Equivariant Imaging: an Unsupervised Framework for Multispectral Pansharpening](https://arxiv.org/abs/2403.09327), as the homography provides a richer group of transformations and so provides better performance for equivariant imaging in camera-based imaging scenarios such as satellite imaging.

See further demos of this functionality [at this webpage notebook](https://andrewwango.github.io/perspective-equivariant-imaging/demo.html#appendix-subgroup-transforms).

Note this adds `kornia` as a dependency, but happy to make it optional dependency instead if needed.

### Checks to be done before submitting your PR
- [x] `python3 -m pytest tests/` runs successfully.
- [x] `black .` runs successfully.
- [x] `make html` runs successfully (in the `docs/` directory).
- [x] Updated docstrings related to the changes (as applicable).
- [x] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

### CODE REVIEW

**Comment by tachella** [comment]
Hey Andrew, 

This would be a nice contribution! I think it would be good to document all of the transforms, and add them in the  [documentation](https://deepinv.github.io/deepinv/deepinv.loss.html#transforms).

I can have a look once the tests are passing!

**Comment by samuro95** [comment]
I am in favor of making the kornia dependency optional, especially because it is used only once, and only in case the transform is applied to tensors. @tachella what do you think ?

**Comment by Andrewwango** [comment]
> I am in favor of making the kornia dependency optional, especially because it is used only once, and only in case the transform is applied to tensors. @tachella what do you think ?

How do I add it as an optional dependency?

**Comment by tachella** [comment]
> > I am in favor of making the kornia dependency optional, especially because it is used only once, and only in case the transform is applied to tensors. @tachella what do you think ?
> 
> How do I add it as an optional dependency?

You can do it in the same way as bm3d for example, see https://github.com/deepinv/deepinv/blob/311f98ee3e30d1a95762b6b8f19d5c536e2e5921/deepinv/models/bm3d.py#L6

File: deepinv/training/testing.py [hunk 1]
```diff
@@ -165,18 +165,13 @@ def test(
...
-            else:
-                save_folder_im = None
-            if plot_metrics:
-                save_folder_curve = (
-                    (save_folder / ("G" + str(g))) if G > 1 else save_folder
-                ) / "curves"
-                save_folder_curve.mkdir(parents=True, exist_ok=True)
-
-            if plot_images:
+                    (
```
Comment (tachella) [inline deepinv/training/testing.py:168]: why did you apply these changes?

File: deepinv/training/trainer.py [hunk 2]
```diff
@@ -717,7 +726,7 @@ def train(
     **kwargs,
 ):
     """
-    Alias function for training a model using :class:`deepinv.training_utils.Trainer` class.
+    Alias function for training a model using :class:`deepinv.training.Trainer` class.
```
Comment (tachella) [inline deepinv/training/trainer.py:729]: I think this should be directly deepinv.Trainer

File: deepinv/transform/projective.py [hunk 3]
```diff
@@ -0,0 +1,330 @@
+from dataclasses import dataclass
+from typing import Union
+
+import numpy as np
+import torch
+from PIL import Image
+
+from scipy.spatial.transform import Rotation
```
Comment (tachella) [inline deepinv/transform/projective.py:8]: Do you think it is possible to remove this dependency?
If I understood well, it is used for generating a rotation matrix from euclidean angles on a single line of code.

File: deepinv/transform/projective.py [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+    Subgroup transformations include :class:`deepinv.transform.Affine`, :class:`deepinv.transform.Similarity`,
+    :class:`deepinv.transform.Euclidean` along with the basic :class:`deepinv.transform.Shift`,
+    :class:`deepinv.transform.Rotation` and semigroup :class:`deepinv.transform.Scale`.
+
+    Transformations with perspective effects (i.e. pan+tilt) are recovered by setting
+    theta_max > 0.
+
+    Generates n_trans random transformations concatenated along the batch dimension.
+
+    Example:
```
Comment (tachella) [inline deepinv/transform/projective.py:143]: the standard way of adding examples is adding first a separator via |sep| 
and then using >>> (otherwise the doctests are not run,  and the automatic copy-paste button will not work) as in [here](https://github.com/deepinv/deepinv/blob/0d0c174f4e5fa9031439bcac82631420c745ed28/deepinv/physics/blur.py#L323)

File: deepinv/transform/projective.py [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+
+    Special case of homography which corresponds to the actions of the affine subgroup
+    Aff(3). Affine transformations include translations, rotations, reflections,
+    skews, and stretches. These transformations are parametrised using geometric parameters in the pinhole camera model. See :class:`deepinv.transform.Homography` for more details.
+
+    Generates n_trans random transformations concatenated along the batch dimension.
+
+    Example:
+
+    ::
```
Comment (tachella) [inline deepinv/transform/projective.py:232]: same here

File: deepinv/transform/projective.py [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+class Similarity(Homography):
+    """Random 2D similarity image transformations.
+
+    Special case of homography which corresponds to the actions of the similarity subgroup
+    S(2). Similarity transformations include translations, rotations, reflections and
+    uniform scale. These transformations are parametrised using geometric parameters in the pinhole camera model. See :class:`deepinv.transform.Homography` for more details.
+
+    Generates n_trans random transformations concatenated along the batch dimension.
+
+    Example:
```
Comment (tachella) [inline deepinv/transform/projective.py:256]: same here

File: deepinv/transform/projective.py [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+
+
+class Similarity(Homography):
+    """Random 2D similarity image transformations.
+
+    Special case of homography which corresponds to the actions of the similarity subgroup
+    S(2). Similarity transformations include translations, rotations, reflections and
+    uniform scale. These transformations are parametrised using geometric parameters in the pinhole camera model. See :class:`deepinv.transform.Homography` for more details.
+
+    Generates n_trans random transformations concatenated along the batch dimension.
```
Comment (tachella) [inline deepinv/transform/projective.py:254]: n_trans -> ``n_trans`` using double `

File: deepinv/transform/projective.py [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+
+
+class Affine(Homography):
+    """Random affine image transformations.
+
+    Special case of homography which corresponds to the actions of the affine subgroup
+    Aff(3). Affine transformations include translations, rotations, reflections,
+    skews, and stretches. These transformations are parametrised using geometric parameters in the pinhole camera model. See :class:`deepinv.transform.Homography` for more details.
+
+    Generates n_trans random transformations concatenated along the batch dimension.
```
Comment (tachella) [inline deepinv/transform/projective.py:228]: n_trans -> ``n_trans`` using double `

File: deepinv/transform/projective.py [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+
+    Special case of homography which corresponds to the actions of the Euclidean subgroup
+    E(2). Euclidean transformations include translations, rotations and reflections. These transformations are parametrised using geometric parameters in the pinhole camera model.
+    See :class:`deepinv.transform.Homography` for more details.
+
+    Generates n_trans random transformations concatenated along the batch dimension.
+
+    Example:
+
+    ::
```
Comment (tachella) [inline deepinv/transform/projective.py:285]: same here, use >>> for doctests and copybutton

File: deepinv/transform/projective.py [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+
+
+class Euclidean(Homography):
+    """Random Euclidean image transformations.
+
+    Special case of homography which corresponds to the actions of the Euclidean subgroup
+    E(2). Euclidean transformations include translations, rotations and reflections. These transformations are parametrised using geometric parameters in the pinhole camera model.
+    See :class:`deepinv.transform.Homography` for more details.
+
+    Generates n_trans random transformations concatenated along the batch dimension.
```
Comment (tachella) [inline deepinv/transform/projective.py:281]: same here with n_trans -> ``n_trans`` using double `

File: deepinv/transform/projective.py [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+    The transformations simulate panning, tilting or rotating the camera, leading to a
+    "perspective" effect. The subgroup is isomorphic to SO(3).
+
+    See :class:`deepinv.transform.Homography` for more details.
+
+    Generates n_trans random transformations concatenated along the batch dimension.
+
+    Example:
+
+    ::
```
Comment (tachella) [inline deepinv/transform/projective.py:317]: Same here with >>>

File: examples/self-supervised-learning/demo_ei_transforms.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+# masked with a random mask, without ground truth, using EI. For this
+# example we use the Urban100 images of natural urban scenes. As these
+# scenes are imaged with a camera free to move and rotate in the world,
+# all of the above transformations are valid invariances that we can
+# impose on the unknown image set :math:`x\in X`.
+#
+
+physics = dinv.physics.Inpainting((3, 256, 256), mask=0.6, device=device)
+
+download_and_extract_archive(
```
Comment (tachella) [inline examples/self-supervised-learning/demo_ei_transforms.py:87]: can you please update to the Urban100HR class added in a recent PR?

File: examples/self-supervised-learning/demo_ei_transforms.py [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+
+train_dataloader, test_dataloader = DataLoader(train_dataset, shuffle=True), DataLoader(
+    test_dataset
+)
+
+
+# %%
+# For training, use a small UNet, Adam optimizer, EI loss with homography
+# transform, and the ``deepinv.Trainer`` functionality:
+#
```
Comment (tachella) [inline examples/self-supervised-learning/demo_ei_transforms.py:109]: maybe add a 

.. note:: 

that clarifies that we only train for a single epoch in the demo, but it is recommended to train x epochs for training the network from scratch.

**Review by tachella — CHANGES_REQUESTED** [review]
The PR looks great, I just found a few details in the docstrings to correct: at the moment the example snippets will not run since the :: indicator for code is not recognized by doctest.

File: deepinv/training/testing.py (hunk 1 above)
Comment (Andrewwango) [reply deepinv/training/testing.py:168]: I believe that was @samuro95 ?

File: deepinv/transform/projective.py (hunk 3 above)
Comment (Andrewwango) [reply deepinv/transform/projective.py:8]: Done!

File: examples/self-supervised-learning/demo_ei_transforms.py (hunk 12 above)
Comment (Andrewwango) [reply examples/self-supervised-learning/demo_ei_transforms.py:87]: This increases demo runtime by ~1min as the current deepinv Urban100HR is much bigger than the HF Urban100 I'm using. I could create an issue?

**Review by tachella — APPROVED** [review]
many thanks @Andrewwango for this PR - everything looks good to me
